keps/sig-node/127-user-namespaces/README.md

```diff
@@ -0,0 +1,456 @@
+capability to grant me any extra privileges on the host or other pods.
+
+#### Story 2
+
+As a cluster admin, I want to allow some pods to run in the host user namespace
+if they need a feature only available in such user namespace, such as loading a
+kernel module with `CAP_SYS_MODULE`.
+
+#### Story 3
+
+As a cluster admin, I want to allow users to run their container as root
+without that process having root privileged on the host, so I can mitigate the
+impact of a compromised container.
+
+#### Story 4
+
+As a cluster admin, I want to allow users to choose the user/group they use
+inside the container without it being the user/group they use on the node, so I
+can mitigate the impact of a container breakout vulnerability (e.g to read
+host files).
+
+#### Story 5
+
+As a cluster admin, I want to use different host UIDs/GIDs for pods running in
+the same node (whenever kernel/kube features allow it), so I can mitigate the
+impact a compromised pod can have on other pods and the node itself.
+
+### Notes/Constraints/Caveats
+
+#### UID space
+
+#### Volume support
+
+### Risks and Mitigations
+
+## Design Details
+
+Note: Names are preliminary yet, I'm using field names to simplify explanations.
+
+### Pod.spec changes
+
+The following changes will be done to the pod.spec:
+
+- `pod.spec.useHostUsers`: bool.
+If true or not present, uses the host user namespace (as today)
+If false, a new userns is created for the pod.
+This field will be used for phase 1, 2 and 3.
+
+- `pod.spec.securityContext.userns.pod2podIsolation`: Enum
+If enabled, we will make the userns mappings be non-overlapping as much as possible.
+This field will be used in phase 3.
+
+### Phases
+
+We propose to divide the work in 3 phases. Each phase makes this work with
+either more isolation or more workloads. When no support is yet added to handle
+some workload, a clear error will be shown.
+
+Please note the last sub-section here is a table with the summary of the changes
+proposed on each phase.
+
+#### Phase 1: pods "without" volumes
+
+This phase makes pods "without" volume work with user namespaces. This is
+activated via the bool `pod.spec.HostUsers` and works only when no volume or any
+combination of the following volume types are used by the pod:
+
+ - configmap
+ - secret
+ - downwardAPI
+ - emptyDir
+ - projected
+
+This list of volumes was chosen as they can't be used to share files with other
+pods.
+
+The mapping length will be 65536, mapping the range 0-65536 to the pod. This wide
+range makes sure most workloads will work fine.
+
+The mapping will be chosen by the kubelet, using a simple algorithm to give
+different pods in this category ("without" volumes) a non-overlapping mapping.
+Giving non-overlapping mappings generates the best isolation for pods.
+
+Furthermore, the node UID space of 2^32 can hold up to 2^16 pods each with a
+mapping length of 65536 (2^16). This imposes a limit of 65536 pods per node, but
```

Comment:
2^32 is the entire space but some IDs will be reserved for the node/host, right?

Like, no pod will map to root (ID 0) on the host, so the available space is 2^32 minus one? Are there any other IDs that are not given out to pods?

Reply:
Well, the range the kubelet will use will be configurable (further PRs will add more details on that).

Yes, 2^32-1 is the complete UID space. The recommendation will (probably) be to reserve the first 2^16 UIDs for the host so it avoids overlapping UIDs used by the host (in process/host files) with UIDs used by pods.

However, the idea is to show that this upper bound doesn't restrict the number of pods we can run in any significant way. I added the word "top", so it says:
> can hold up to 2^16 pods each with a mapping length of 65535 (2^16-1) *top*.

I think it is clear it doesn't, even if the adm decides to reserve some ranges of 16 bits.